### PR TITLE
Sub-slice with worker group annotation

### PR DESCRIPTION
**Pull request by spencer-p**
This change implements sub-slicing via a worker group annotation.

1. The user specifies a node selector as normal, but this now references the "parent" slice.
2. The user adds an annotation specifying the desired subslice topology.
3. The webhook uses the subslice topology annotation to verify the numOfHosts, if present, instead of the parent slice.
4. The webhook verifies there is a node label that can be used to guarantee scheduling (topology host, subblock, block, or the full nodepool)
5. The mutate webhook injects a more narrow affinity to schedule on the subslice if needed.
6. The device plugin will consume the annotation and set the correct env vars.

Example YAML snippet:

```yaml
  workerGroupSpecs:
  - groupName: tpu-group
    replicas: 1
    maxReplicas: 1
    numOfHosts: 2  # Matches 2x4 slice
    groupName: tpu-group
    rayStartParams: {}
    template:
      metadata:
        annotations:
          cloud.google.com/gke-tpu-slice-topology: 2x4  # !!! New !!!
      spec:
        nodeSelector:
          cloud.google.com/gke-tpu-accelerator: tpu-v6e-slice
          cloud.google.com/gke-tpu-topology: 4x4  # Parent slice shape
        containers:
        - name: ray-worker
```

To discuss
- ~~Should we use a different mechanism than an annotation? The nodeselector can't be used, but we don't _have_ to use annotations.~~
   - We need this specific annotation to interface with the device plugin
- Should we reject clusters if the TPUs are allocated?
   - No, that's too advanced, and we do need to admit the cluster when no nodes are available to trigger scale-up.
- This only works if pre-provisioned nodes exist. Theoretically the mapping from (tpu type, parent slice, subslice) -> affinity rule is fixed and could be statically compiled in instead of determined at runtime. That would require testing on all slice sizes for all TPU types (!!!) but would be much better than listing nodes.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -348,11 +359,35 @@ func makeLabelSelectorRequirement(key string, op metav1.LabelSelectorOperator, v
...
 // injectAffinity injects pod affinity and anti-affinity scheduling constraints using replicaIndex and cluster labels
 // to ensure TPU Pods from the same multi-host replica are co-located.
-func injectAffinity(pod *corev1.Pod, replicaIndex int, workerGroupName string, patches *[]patch) {
+func (t *TPUWebhookServer) injectAffinity(pod *corev1.Pod, replicaIndex int, numOfHosts int, workerGroupName string, patches *[]patch) {
 	clusterName := pod.Labels[utils.RayClusterLabelKey]
-	topologyKey := "cloud.google.com/gke-nodepool"
+	topologyKey := getGKETopologyKey(pod)
+
+	// If the current topology key is the default nodepool, attempt to discover a more specific one (block/subblock).
+	if topologyKey == "cloud.google.com/gke-nodepool" && t != nil && t.nodeLister != nil {
```
Comment (ryanaoleary) [inline main.go]: Can we update this section in the README: https://github.com/ai-on-gke/kuberay-tpu-webhook#what-the-webhook-does-automatically so it's clear to users how to use the subslicing feature with the webhook

Comment (spencer-p) [reply main.go]: Done.

File: main.go [hunk 2]
```diff
@@ -348,11 +361,45 @@ func makeLabelSelectorRequirement(key string, op metav1.LabelSelectorOperator, v
 	}
 }
 
+// getGKETopologyKey returns the first GKE topology key found in the Pod's affinity,
+// defaulting to the nodepool key if none are found.
+func getGKETopologyKey(pod *corev1.Pod) string {
+	if pod.Spec.Affinity != nil && pod.Spec.Affinity.PodAffinity != nil {
+		for _, term := range pod.Spec.Affinity.PodAffinity.RequiredDuringSchedulingIgnoredDuringExecution {
+			if strings.HasPrefix(term.TopologyKey, "cloud.google.com/") {
```
Comment (ryanaoleary) [inline main.go]: nit: could make "cloud.google.com/" a constant

File: main.go [hunk 3]
```diff
@@ -348,11 +361,45 @@ func makeLabelSelectorRequirement(key string, op metav1.LabelSelectorOperator, v
...
+				return term.TopologyKey
+			}
+		}
+	}
+	return "cloud.google.com/gke-nodepool"
```
Comment (ryanaoleary) [inline main.go]: `return gkeNodePoolLabel`

File: main.go [hunk 4]
```diff
@@ -1264,25 +1479,41 @@ func main() {
 	factory := informers.NewFilteredSharedInformerFactory(client, 1*time.Minute, metav1.NamespaceAll, tweakListOptionsFunc)
 	podInformer := factory.Core().V1().Pods().Informer()
 
-	// start the PodInformer and wait for cache sync
+	// instantiate NodeInformer for TPU nodes in the GKE cluster
+	tweakNodeListOptionsFunc := func(options *metav1.ListOptions) {
+		options.LabelSelector = "cloud.google.com/gke-tpu-accelerator"
```
Comment (ryanaoleary) [inline main.go]: nit: add `cloud.google.com/gke-tpu-accelerator` as a constant as well and then update the assignment here / where it's used in tests

File: README.md [hunk 5]
```diff
@@ -82,6 +83,7 @@ The webhook evaluates each `workerGroupSpec` against the following rules:
     * **Expected Hosts:** `max(Total Chips / Chips Per Host, 1)`
     * If the calculated `Expected Hosts` does not exactly match the `numOfHosts` defined in your `workerGroupSpec`, the cluster is rejected with the error: `"Number of workers in worker group not equal to specified topology"`.
   * **Example:** If your node selector specifies a `2x2x2` topology (8 total chips) and your container requests `4` TPUs (`google.com/tpu: "4"`), your `numOfHosts` must be set to `2`.
+* **Subslice Ambiguity Validation:** If a subslice is requested via the `cloud.google.com/gke-tpu-slice-topology` annotation, the webhook verifies that the requested `numOfHosts` corresponds to a unique physical grouping (block, subblock, etc.) available in the targeted nodes. If the request is ambiguous or no grouping matches the host count, the cluster is rejected.
```
Comment (ryanaoleary) [inline README.md]: can we add a sentence mentioning how the raycluster may be admitted with a warning in order to trigger scaleup for Autopilot - but that subslice scheduling will not work in this case

**Comment by ryanaoleary** [comment]
minor comments but mostly LGTM, will run some manual tests on this PR (I'm interested if the nodeLister adds any noticeable latency) and approve/merge this monday dependent on that

File: main.go (hunk 2 above)
Comment (spencer-p) [reply main.go]: Done.

File: main.go (hunk 3 above)
Comment (spencer-p) [reply main.go]: Done.

File: main.go (hunk 4 above)
Comment (spencer-p) [reply main.go]: Done.

File: README.md (hunk 5 above)
Comment (spencer-p) [reply README.md]: Done.

File: README.md [hunk 6]
```diff
@@ -67,6 +67,7 @@ When you submit a RayCluster resource requesting TPUs, this mutating webhook int
...
 * **Megascale (Multi-Slice) Support:** If `MEGASCALE_NUM_SLICES` is set explicitly in the Pod spec of your Ray container, the webhook automatically calculates and injects `MEGASCALE_SLICE_ID`, `MEGASCALE_COORDINATOR_ADDRESS`, and `MEGASCALE_PORT`. If utilizing the [JaxTrainer](https://docs.ray.io/en/latest/train/api/doc/ray.train.v2.jax.JaxTrainer.html#ray.train.v2.jax.JaxTrainer) in Ray Train, `MEGASCALE_NUM_SLICES` and related env vars are calculated for you based on the value of `num_workers`, `accelerator_type`, and `topology` and set automatically at runtime.
+* **TPU Subslicing Support:** If `cloud.google.com/gke-tpu-slice-topology` is set as an annotation on the Pod template, the webhook automatically calculates and injects the appropriate `nodeAffinity` (e.g., `cloud.google.com/gce-topology-subblock`) to ensure that all hosts in a subslice are co-located on the same physical grouping within a larger TPU slice. This enables running multiple independent smaller workloads on a single large TPU reservation.
```
Comment (ryanaoleary) [inline README.md]: I don't think we're actually injecting a `nodeAffinity` here right? just pod affinity / anti-affinity

Comment (ryanaoleary) [reply README.md]: i.e. I don't think we're using the https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/ node affinity field

File: main.go [hunk 7]
```diff
@@ -348,11 +367,45 @@ func makeLabelSelectorRequirement(key string, op metav1.LabelSelectorOperator, v
...
-	topologyKey := "cloud.google.com/gke-nodepool"
+	topologyKey := getGKETopologyKey(pod)
+
+	// If the current topology key is the default nodepool, attempt to discover a more specific one (block/subblock).
+	if _, subsliceRequested := pod.Annotations[tpuSubsliceTopologyAnnotation]; subsliceRequested {
+		selector := labels.SelectorFromSet(pod.Spec.NodeSelector)
+		nodes, err := t.nodeLister.List(selector)
+		if err == nil && len(nodes) > 0 {
+			topology := buildTPUTopology(nodes)
+			if key, err := topology.subsliceAffinityKey(numOfHosts); err == nil {
```
Comment (ryanaoleary) [inline main.go:395]: I think we're missing the handling if `err != nil` here and/or topology key is empty

File: README.md (hunk 6 above)
Comment (spencer-p) [reply README.md]: Thanks, yeah that's not right.

File: main.go (hunk 7 above)
Comment (spencer-p) [reply main.go:395]: Good catch. Fixed. Will add a test too

File: main.go [hunk 8]
```diff
@@ -1203,6 +1341,93 @@ func (c *podSyncCond) Signal(pod string) bool {
...
+// TPUTopology indexes the statically discoverable topology labels for TPU
+// nodes.
+type TPUTopology struct {
+	Hosts     map[string][]*corev1.Node
+	Subblocks map[string][]*corev1.Node
+	Blocks    map[string][]*corev1.Node
+	NodePools map[string][]*corev1.Node
+}
+
+func buildTPUTopology(nodes []*corev1.Node) *TPUTopology {
```
Comment (ryanaoleary) [inline main.go:1385]: should we ever be checking if `node.Spec.Unschedulable == true` before using its labels here?

**Review by ryanaoleary — APPROVED** [review]
one last comment but its non-blocking, LGTM!

File: main.go (hunk 8 above)
Comment (ryanaoleary) [reply main.go:1385]: my intuition says no because the nodepool could be repaired, etc. but just wondering

Comment (spencer-p) [reply main.go:1385]: I thought about this (as well as checking whether there is already a pod consuming the TPUs). But I agree, I don't think we should validate at that level, and leave repairing and scheduling to the other components.
